RC Refactor + (some) BT-encode fixes

dr7ana opened the pull request:
Picking up on RC/BT work from #2217

dr7ana commented:
Accidentally closed when the commit was dropped, re-opening tomorrow with a re-working of routercontacts

dr7ana committed:
RC refactor layout

- Local and Remote RC's now implemented with discrete functionalities and uses

jagerman commented on llarp/router_contact_remote.cpp:
```diff
@@ -0,0 +1,69 @@
+
+    // TODO: implement append_encoded in oxenc so we can finish this implementation.
+    // It is almost identical to the implementation of LocalRC::bt_encode, except the
+    // call to append_signature is append_encoded.
```
As described this won't work, because it will potentially break signatures (if the RC we received had extra keys we didn't understand, from a future Lokinet; or ones we ignore from a past lokinet).  We'll have to instead get the entire, untouched, encoded RC as we received it and push that out.

jagerman replied:
For example, future lokinet might decide it doesn't care about ipv6 anymore and stops parsing the ipv6 field entirely, so then this:
```json
{"4": "abcdpp", "6": "whateverpp", "p": "...pubkey...", "~": "xyz..."}
```
would get round-trip re-encoded as:
```json
{"4": "abcdpp", "p": "...pubkey...", "~": "xyz..."}
```
but now the `xyz...` won't match anymore because the "6" will be missing from the serialized dict output.

dr7ana replied:
I know it wont work! I remember your explanation on that. To clarify, I meant we need `append_encoded` for adding the signature to the end when we call `RemoteRC::bt_encode`, not the individual key/value pairs. For `LocalRC::bt_encode`, we will need `append_signature` (which we already have) at the very end. I can explain more over mumble when you're online

jagerman commented on llarp/crypto/types.cpp:
```diff
@@ -32,14 +33,44 @@ namespace llarp
+  operator==(const RouterID& lhs, const PubKey& rhs)
+  {
+    return lhs.as_array() == rhs.as_array();
+  }
```
I'm pretty sure that we can get rid of both of these from here and the `.hpp` because the two-PubKey `operator==` should already properly match any pubkey+routerid pair (due to the added inheritance).

jagerman commented on llarp/crypto/types.cpp:
```diff
@@ -32,14 +33,44 @@ namespace llarp
+  }
+
   bool
   SecretKey::LoadFromFile(const fs::path& fname)
```
I feel like we should get rid of these `LoadFromFile` methods and just have a single function something like:
```C++
template <typename T, typename = std::enable_if_t<is_aligned_buffer<T>>>
T load_from_file(const fs::path& fname) {
    ...
}

// ...
auto sk = load_from_file<SecretKey>("filename");
```
Having these nearly-identical `LoadFromFile` scatter all over the codebase (making `fs.hpp` included almost everywhere) seems sub-optimal.

dr7ana replied:
I agree, but question and a note:

Q: What other `LoadFromFile` methods are you referring to? I only see the one in `SecretKey` used only in two places
```
dan@demeter:~/oxen/lokinet/llarp$ grep -R "LoadFromFile" *
config/key_manager.cpp:    return key.LoadFromFile(path);
crypto/types.hpp:    LoadFromFile(const fs::path& fname);
crypto/types.cpp:  SecretKey::LoadFromFile(const fs::path& fname)
exit/session.cpp:    return exit_key.LoadFromFile(fname);
```

Note: We already have other free methods for reading files directly to buffers or string [here](https://github.com/oxen-io/lokinet/blob/07271f9ae7a9d0f39c3b866d06d625e218315419/llarp/util/file.hpp#L13). Their former name is absolutely terrible so I re-named them, but maybe we can modify these to be more what you're thinking of?

dr7ana committed:
{Remote,Local}RC's

- RemoteRC supplants most of the functionality throughout the code of RouterContact
- Next step will be to sort out CI issues, then see if we can get rid of either LocalRC (and therefore RouterContact entirely)

dr7ana committed:
Review fixes + misc fixes

jagerman committed:
Untangle Endpoint::LookupServiceAsync

- .snodes don't need to support SRV records, so remove that
- untangle the mess of captured lambdas capturing other lambdas
  capturing other lambdas; we still need a chain of nested lambdas
  because we have a chain of callbacked events, but hiding the nesting
  by capturing them in other lambdas didn't improve anything.

tewinget commented on llarp/crypto/types.hpp:
```diff
@@ -161,58 +143,24 @@ namespace llarp
   using ShortHash = AlignedBuffer<SHORTHASHSIZE>;
   using LongHash = AlignedBuffer<HASHSIZE>;
 
   struct Signature final : public AlignedBuffer<SIGSIZE>
```
`using`?

jagerman commented on llarp/crypto/types.cpp:
```diff
@@ -32,14 +33,38 @@ namespace llarp
+  operator==(const PubKey& lhs, const RouterID& rhs)
+  {
+    return lhs.as_array() == rhs.as_array();
+  }
```
This one can go away too.

jagerman commented on llarp/crypto/types.cpp:
```diff
@@ -32,14 +33,38 @@ namespace llarp
+  PubKey::operator RouterID() const
+  {
+    return {as_array()};
+  }
```
I wonder if this is still needed?
